### PR TITLE
DOC/TYP: index.take return val

### PR DESCRIPTION
**Pull request by mzeitlin11**
xref https://github.com/pandas-dev/pandas/pull/40513#discussion_r597541594

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -907,16 +907,26 @@ def astype(self, dtype, copy=True):
...
     @Appender(_index_shared_docs["take"] % _index_doc_kwargs)
-    def take(self, indices, axis=0, allow_fill=True, fill_value=None, **kwargs):
+    def take(
+        self,
+        indices: Union[ArrayLike, Sequence[int]],
+        axis: int = 0,
+        allow_fill: bool = True,
+        fill_value=None,
+        **kwargs,
+    ):
```
Comment (mzeitlin11) [inline pandas/core/indexes/base.py:917]: Any advice on typing this return value? Can almost use the TypeVar `_IndexT`, but the problem is that `RangeIndex.take` returns a superclass (`Int64Index`).

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -907,16 +907,26 @@ def astype(self, dtype, copy=True):
         """
 
     @Appender(_index_shared_docs["take"] % _index_doc_kwargs)
-    def take(self, indices, axis=0, allow_fill=True, fill_value=None, **kwargs):
+    def take(
+        self,
+        indices: Union[ArrayLike, Sequence[int]],
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:912]: is ArrayLike really needed here?  it should also be a Sequence[int]

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -886,7 +886,7 @@ def astype(self, dtype, copy=True):
 
         Parameters
         ----------
-        indices : list
+        indices : array_like
```
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py:889]:
```suggestion
        indices : array-like
```

https://pandas.pydata.org/pandas-docs/dev/development/contributing_docstring.html#parameter-types

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py:912]:
```suggestion
        indices: Union[AnyArrayLike, Sequence[int]],
```

can also pass a Series or Index to `Index.take`

Comment (simonjayhawkins) [reply pandas/core/indexes/base.py:912]: see #28770 for issue about EA, but applicable to Series, Index and np.ndarray also

```
from typing import Sequence, cast
import numpy as np
import pandas as pd
from pandas.core.arrays import ExtensionArray

df = pd.DataFrame({"a": [1, 2, 3]})
idx = df.index
arr: np.ndarray = cast(np.ndarray, idx._values)
ea: ExtensionArray = pd.array(arr, dtype="Int64")
ser = pd.Series(arr)
reveal_type(idx)
reveal_type(arr)
reveal_type(ea)
reveal_type(ser)


def func(arr: Sequence):
    pass


func(idx)
func(arr)
func(ea)
func(ser)
```
```
/home/simon/t.py:11: note: Revealed type is 'pandas.core.indexes.base.Index'
/home/simon/t.py:12: note: Revealed type is 'numpy.ndarray'
/home/simon/t.py:13: note: Revealed type is 'pandas.core.arrays.base.ExtensionArray'
/home/simon/t.py:14: note: Revealed type is 'pandas.core.series.Series'
/home/simon/t.py:21: error: Argument 1 to "func" has incompatible type "Index"; expected "Sequence[Any]"  [arg-type]
/home/simon/t.py:22: error: Argument 1 to "func" has incompatible type "ndarray"; expected "Sequence[Any]"  [arg-type]
/home/simon/t.py:23: error: Argument 1 to "func" has incompatible type "ExtensionArray"; expected "Sequence[Any]"  [arg-type]
/home/simon/t.py:24: error: Argument 1 to "func" has incompatible type "Series"; expected "Sequence[Any]"  [arg-type]
```

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/indexes/base.py:917]: I _think_ in the base class use `Index`, and can then use a typevar in the subclasses that do return the same type.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @mzeitlin11 for the PR. generally lgtm, a couple of suggestions.

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (mzeitlin11) [reply pandas/core/indexes/base.py:912]: Thanks, makes sense

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (mzeitlin11) [reply pandas/core/indexes/base.py:917]: I tried something along those lines originally - the problem I was running into is how to handle subclasses which rely on parent implementations. For example, for `ExtensionIndex` would like to be able to write a stub like
```
def take(
        self: _ExtensionIndexT,
        indices,
        axis: int = 0,
        allow_fill: bool = True,
        fill_value=None,
        **kwargs,
    ) -> _ExtensionIndexT: ...
```
to essentially type the child while still relying on the inherited implementation. But couldn't figure out a way to achieve this.

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -907,16 +907,26 @@ def astype(self, dtype, copy=True):
         """
 
     @Appender(_index_shared_docs["take"] % _index_doc_kwargs)
-    def take(self, indices, axis=0, allow_fill=True, fill_value=None, **kwargs):
+    def take(
+        self,
+        indices: Union[AnyArrayLike, Sequence[int]],
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:912]: isnt Sequence[int] going to cover the AnyArrayLike case?

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -886,7 +886,7 @@ def astype(self, dtype, copy=True):
 
         Parameters
         ----------
-        indices : list
+        indices : array-like
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:889]: i think Sequence[int] would be more helpful

File: pandas/core/indexes/base.py (hunk 4 above)
Comment (mzeitlin11) [reply pandas/core/indexes/base.py:912]: I think even with `AnyArrayLike` in place of `ArrayLike` we run into the issue mentioned here https://github.com/pandas-dev/pandas/pull/40521#discussion_r598605341

File: pandas/core/indexes/base.py (hunk 5 above)
Comment (mzeitlin11) [reply pandas/core/indexes/base.py:889]: Happy to change, my only question would be if we want to be using type hints for user facing API. The numpy documentation for `take` just uses `array_like`

File: pandas/core/indexes/base.py (hunk 4 above)
Comment (jreback) [reply pandas/core/indexes/base.py:913]: this is too complicated and specific an annotation to put *here*, leave this untyped for now, or put it in a typing alias, but then we should enforce it, alt we shouldn't allow non-array-likes here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 4 above)
Comment (mzeitlin11) [reply pandas/core/indexes/base.py]: Left untyped for now, now mostly just a doc fix

File: pandas/core/indexes/base.py (hunk 5 above)
Comment (simonjayhawkins) [reply pandas/core/indexes/base.py:934]: > i think Sequence[int] would be more helpful

we shouldn't use typing notation for docstrings. https://pandas.pydata.org/pandas-docs/dev/development/contributing_docstring.html#parameter-types

and from that guide...

> If the exact type is not relevant, but must be compatible with a NumPy array, array-like can be specified. If Any type that can be iterated is accepted, iterable can be used:

but the guide doesn't mention sequence. 

maybe `sequence of int` is allowed, but I think matching numpy is fine.

**Comment by simonjayhawkins** [comment]
@jbrockmendel can you take another look, @jreback comments have been addressed.

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by simonjayhawkins** [comment]
Thanks @mzeitlin11

**Pull request merged**